torch/csrc/jit/codegen/cuda/lower_trivial_broadcast.cpp

```diff
@@ -0,0 +1,176 @@
+    }
+  }
+  traverse(fusion);
+}
+
+bool ConcretizedBroadcastDomains::isConcretized(IterDomain* id) const {
+  auto it = concretized_domains_.find(id);
+  return it != concretized_domains_.end();
+}
+
+void ConcretizedBroadcastDomains::handle(BroadcastOp* bop) {
+  // Create a new entry for each of new broadcast domains
+  auto out = bop->out()->as<TensorView>();
+  for (const auto i : c10::irange(out->getRootDomain().size())) {
+    if (bop->getBroadcastDimFlags().at(i)) {
+      auto new_bcast_id = out->getRootDomain().at(i);
+      broadcast_origin_map_.emplace(
+          new_bcast_id, std::unordered_set<IterDomain*>({new_bcast_id}));
+    }
+  }
+}
+
+void ConcretizedBroadcastDomains::handle(Expr* expr) {
+  IterVisitor::handle(expr);
+
+  // Propagate broadcast origin info from producers to consumers
+  for (auto producer : ir_utils::filterByType<TensorView>(expr->inputs())) {
+    std::unordered_set<IterDomain*> producer_broadcasts;
+    for (auto producer_id : producer->getMaybeRFactorDomain()) {
+      if (producer_id->isBroadcast()) {
+        producer_broadcasts.insert(producer_id);
+      }
+    }
+    if (producer_broadcasts.empty()) {
+      continue;
+    }
+
+    for (auto consumer : ir_utils::filterByType<TensorView>(expr->outputs())) {
+      auto p2c_map =
+          PairwiseRootDomainMap(producer, consumer)
+              .mapProducerToConsumer(
+                  producer->domain(), consumer->domain(), producer_broadcasts);
+      for (const auto& kv : p2c_map) {
+        auto p_id = kv.first;
+        auto c_id = kv.second;
+        const bool is_concretized = !c_id->isBroadcast();
+        auto it = broadcast_origin_map_.find(p_id);
+        TORCH_INTERNAL_ASSERT(
+            it != broadcast_origin_map_.end(),
+            "Broadcast origin info not found for producer broadcast domain: ",
+            p_id->toString(),
+            " of ",
+            producer->toString());
+        const auto& producer_origins = it->second;
+        if (is_concretized) {
+          // Keep track of all the origin domains as concretized
+          for (auto origin : producer_origins) {
+            // concretized_root_domains_.insert(origin);
+            markAsConcretized(origin);
+          }
+        } else {
+          // Not concretized yet. Propagate forward the origin info.
+          auto& consumer_origins = broadcast_origin_map_[c_id];
+          for (auto origin : producer_origins) {
+            consumer_origins.insert(origin);
+          }
+          consumer_origins.insert(c_id);
+        }
+      }
+    }
+  }
+}
+
+void ConcretizedBroadcastDomains::markAsConcretized(IterDomain* root_domain) {
+  concretized_domains_.emplace(root_domain);
+  std::deque<IterDomain*> child_domains({root_domain});
+  while (!child_domains.empty()) {
+    auto child = child_domains.front();
+    child_domains.pop_front();
+    const auto& child_uses = child->uses();
+    if (child_uses.size() == 1 && !child_uses[0]->outputs().empty() &&
```

Comment:
Why not just go through all uses and mark every iter domain as concretized?
Couldn't we even use `DependencyChain::getAllDependentVals` to do this?

Reply:
`DependencyChain::getAllDependentVals` goes in the wrong direction, and the use versions in dependency check just do fusion traversal which starts at terminating outputs, so we'd have to enable the full recursive version of iter visitor which I'm not really ready to do because of perf impact.

Reply:
I made this more aggressive in the last commit I pushed, if you don't like this or think it's wrong feel free to revert. I didn't think this condition was important, but I might be wrong as I didn't understand why.